stratosphere/sm/source/sm_service_manager.cpp

```diff
@@ -0,0 +1,723 @@
+
+            /* Debug. */
+            u64 max_sessions;
+            bool is_light;
+
+            /* Mitm Extension. */
+            u64 mitm_pid;
+            Handle mitm_port_h;
+            Handle mitm_query_h;
+
+            /* Acknowledgement members. */
+            bool mitm_waiting_ack;
+            u64 mitm_waiting_ack_pid;
+            Handle mitm_fwd_sess_h;
+        };
+
+        class AccessControlEntry {
+            private:
+                const u8 *entry;
+                size_t size;
+            public:
+                AccessControlEntry(const void *e, size_t sz) : entry(reinterpret_cast<const u8 *>(e)), size(sz) {
+                    /* ... */
+                }
+
+                AccessControlEntry GetNextEntry() const {
+                    return AccessControlEntry(this->entry + this->GetSize(), this->size - this->GetSize());
+                }
+
+                size_t GetSize() const {
+                    return this->GetServiceNameSize() + 1;
+                }
+
+                size_t GetServiceNameSize() const {
+                    return (this->entry[0] & 7) + 1;
+                }
+
+                ServiceName GetServiceName() const {
+                    return ServiceName::Encode(reinterpret_cast<const char *>(this->entry + 1), this->GetServiceNameSize());
+                }
+
+                bool IsHost() const {
+                    return (this->entry[0] & 0x80) != 0;
+                }
+
+                bool IsWildcard() const {
+                    return this->entry[this->GetServiceNameSize()] == '*';
+                }
+
+                bool IsValid() const {
+                    /* Validate that we can access data. */
+                    if (this->entry == nullptr || this->size == 0) {
+                        return false;
+                    }
+
+                    /* Validate that the size is correct. */
+                    return this->GetSize() <= this->size;
+                }
+        };
+
+        class InitialProcessIdLimits {
+            public:
+                static constexpr u64 InitialProcessIdMin = 0x00;
+                static constexpr u64 InitialProcessIdMax = 0x50;
+            private:
+                u64 min;
+                u64 max;
+            public:
+                InitialProcessIdLimits() {
+                    if (GetRuntimeFirmwareVersion() >= FirmwareVersion_500) {
+                        /* On 5.0.0+, we can get precise limits from the kernel. */
+                        R_ASSERT(svcGetSystemInfo(&this->min, 2, 0, 0));
+                        R_ASSERT(svcGetSystemInfo(&this->max, 2, 0, 1));
+                    } else {
+                        /* On < 5.0.0, we just use hardcoded extents. */
+                        this->min = InitialProcessIdMin;
+                        this->max = InitialProcessIdMax;
+                    }
+
+                    /* Ensure range is sane. */
+                    if (this->min > this->max) {
```

Comment:
They do, however PID was allowed to be zero on lower firmwares, and they added that assert only later. We can't assert that min > 0 because min might be == 0 on lower firmwares.